LUCENE-10062: Switch to numeric doc values for encoding taxonomy ordinals (instead of custom binary format)

gsmiller opened the pull request:
# Description

In benchmarks, using numeric doc values to store taxonomy facet ordinals shows almost a 400% qps improvement in browse-related taxonomy-based tasks (instead of custom delta-encoding into a binary doc values field).

# Solution

This change moves to standard numeric doc values for storing taxonomy ordinals.

# Tests

No new tests added. Lots of existing test coverage for taxonomy faceting functionality.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/lucene/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Lucene maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [ ] I have added tests for my changes.

gsmiller committed:
LUCENE-10062: Switch to numeric doc values for encoding taxonomy ordinals (instead of custom binary format)

mikemccand reviewed (commented):
Wow, what an incredible speedup!  This shows how awesome our default `Codec` has become at storing multiple ints for one field X doc.  Yay :)

But I'm afraid we need some backwards compatibility here.  E.g. say we backport this, and it's released in 8.10.  When a Lucene user upgrades to 8.10, that version must be able to read/write their taxo index as well, do faceting, etc.

We should first add a simple back compat test case, like the one @gautamworah96 added recently on switching from stored fields to doc values in the taxonomy index (LUCENE-9450).  This test unzips an "old" simple Lucene index and confirms we can add new `FacetLabel`s, compute facet counts, etc.  The test should hopefully fail now with this PR (since we don't have back-compat layer yet).  Then add the back-compat layer and test should pass!

jpountz reviewed (commented):
I'm not familiar with the faceting code, I only had a quick look because the 400% speedup caught my attention. :) I like that this makes the faceting code not only faster, but also simpler.

gsmiller committed:
update TaxonomyFacetLabels and TaxonomyFacetSumValueSource to directly access the dv field when possible instead of going through the indirection of OrdinalsReader

gsmiller committed:
Merge branch 'main' into LUCENE-10062-taxo-facet-opto-pr

gsmiller committed:
attempt at making things backwards compatible

gsmiller committed:
Merge remote-tracking branch 'origin/main' into LUCENE-10062-taxo-facet-opto-pr

gsmiller committed:
doc cleanup

gsmiller committed:
support specifying index field in TaxonomyFacetSumValueSource

gsmiller commented on lucene/facet/src/java/org/apache/lucene/facet/taxonomy/DocValuesOrdinalsReader.java:
```diff
@@ -59,16 +53,21 @@ public void get(int docID, IntsRef ordinals) throws IOException {
+          }
+
+          for (int i = 0; i < count; i++) {
+            ordinals.ints[ordinals.length] = (int) dv.nextValue();
```
We're effectively doing this check already in `FacetsConfig` when we write the ordinals since we're explicitly using `int` and `int[]` to track the ordinals for each document before adding them to the `SortedNumericDocValuesField`. So we're always writing `int` values into the field when done through `FacetsConfig`. On the other hand, if the user just goes and indexes whatever they want into a `SortedNumericDocValuesField` and then tries to facet on it, they could run into issues (but that's true in many places in `facets` today).

gsmiller commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -410,7 +411,16 @@ private void processFacetFields(
+      for (int i = 0; i < o.length; i++) {
+        int ord = o.ints[o.offset + i];
+        if (ord > prev) {
+          doc.add(new SortedNumericDocValuesField(indexFieldName, ord));
```
We need to de-dupe because we might add duplicate ordinals to each doc when traversing the ancestry and adding parent docs (for cases where we explicitly index the parents). We could special-case this, but I'm not sure it's worth it given this is during indexing and not during a hot query path? If you feel strongly about it though, I think we can special-case and only do the de-duping when it's necessary.

gsmiller commented on lucene/facet/src/java/org/apache/lucene/facet/taxonomy/DocValuesOrdinalsReader.java:
```diff
@@ -41,12 +40,7 @@ public DocValuesOrdinalsReader(String field) {
 
   @Override
   public OrdinalsSegmentReader getReader(LeafReaderContext context) throws IOException {
```
I ended up keeping support for OrdinalsReader so that users can provide their own source of ordinals in custom cases, but have moved all the default faceting logic over to use the doc values field directly instead of this OrdinalsReader indirection (except in a couple cases to maintain back-compat). So I think I've gotten at the core issue/suggestion here, but certainly open to more suggestions or feeback.

gsmiller replied:
In the latest iteration I've removed `OrdinalsReader` (and support for it) completely. I think this is a nice simplification and I can't think of any real-world cases for supporting a custom `OrdinalsReader` when faceting after we remove support for users bringing their own custom binary encoding. If we come up with real cases for continuing to support this, it's easy enough to bring back.

gsmiller commented:
I've just updated the PR with an attempt at maintaining backwards compatibility with older indexes. So any index created with Lucene 10 or later will use the new numeric doc values format, but indexes created with Lucene 10 or earlier will use the binary format. If the index already exists and was created with Lucene 10 or earlier, the binary format will continue to be used (even if force merging). To get the new format, users will have to re-create the index.

My proposal (which you'll see in the `TODO`s and `@Deprecated` annotations) would be to use the new format in Lucene 10 and beyond, and remove support for the older format when we get to Lucene 11.

gsmiller commented on lucene/facet/src/java/org/apache/lucene/facet/taxonomy/TaxonomyFacetLabels.java:
```diff
@@ -62,7 +62,17 @@ public TaxonomyFacetLabels(TaxonomyReader taxoReader, String indexFieldName) thr
+    // formats for now:
+    // TODO: Remove in Lucene 11
+    if (FacetUtils.usesOlderBinaryOrdinals(readerContext.reader(), indexFieldName)) {
+      OrdinalsReader ordsReader = new DocValuesOrdinalsReader(indexFieldName);
```
We could probably improve this by only creating the OrdinalsReader once and "caching" it, although we'd need to synchronize (or just best effort reuse).

gsmiller committed:
Merge remote-tracking branch 'origin/main' into LUCENE-10062-taxo-facet-opto-pr

rmuir commented on lucene/facet/src/java/org/apache/lucene/facet/FacetUtils.java:
```diff
@@ -81,4 +84,19 @@ public long cost() {
+   */
+  // TODO: Remove in Lucene 11
+  @Deprecated
+  public static boolean usesOlderBinaryOrdinals(LeafReader reader, String field) {
```
Why this check? Can't we simply do `reader.getMetaData().getCreatedVersionMajor() < 9` or whatever, matching the exact logic at write-time in DirectoryTaxonomyWriter?

gsmiller replied:
Good point. Since I moved the back-compat support to #443 (against 9.0) I included this change over there. There's no more back-compat support needed here since this is against main. thanks!

gsmiller commented:
Please note I've opened a separate PR (#443) to add this change to 9.0. After iterating on that PR and merging to 9.0/9.x, I'll come back to this PR and remove the back-compat support (since it will no longer be needed on main if we release this in 9.0). I'm waiting to update this PR though incase significant feedback comes in on #443 (so I can make sure it's included here as well). Apologies if this creates any confusion.

gsmiller committed:
we decided to include this change in 9.0, so there's no need for back-compat on main; updating to reflect this

gsmiller commented:
I've removed all back-compat support now from this PR since we're trying to include this change in 9.0 (see #443). So this PR now reflects my proposed end-state in main. I still suggest we refrain from reviewing this PR in detail until we resolve #443, since any fundamental changes to that PR might require updates here. I've updated though just for reference.

gsmiller committed:
pick up pr feedback including migration guide

msokolov commented:
>  I still suggest we refrain from reviewing this PR in detail

oops, I only read that at the end; still and all I don't have much to say; this area of the code is fairly new for me. Looks like a great simplification and the speedups you found are awesome! On a related topic, while I was stumbling around in here I found we have the option to accumulate scores in MatchingDocs which I'd be excited to explore for doing relevance-weighted sums.

gsmiller committed:
pick up a couple small changes from the 9.0 pr

gsmiller commented:
> On a related topic, while I was stumbling around in here I found we have the option to accumulate scores in MatchingDocs which I'd be excited to explore for doing relevance-weighted sums.

+1, yeah I think it could be a good fit for that use-case. It's a nice feature to be able to "bring your own" scores to accumulate for each doc. One thing you can't do today is accumulate faceting "weights" based on some custom score while also keeping track of counts. It would be nice to be able to, for example, get the top-n values for a particular dim by some notion of relevance while also being able to show how many docs exist for each value. Could be a nice addition.

gsmiller commented:
This is now ready for review. This is the "forward port" of the change introduced to 9.0 on #443 (but without all the back-compat complexity for 8.x indexes).

rmuir reviewed (approved):
Lets get this change in main, as it may create problems that its already in 9.0 and 9.x branches and still not here (e.g. merging difficulties and confusion).
